scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -436,112 +445,232 @@
-                                                       task-info
-                                                       [:executor :executor-id]
-                                                       (:task-id task-info))
-                               :instance/slave-id (:slave-id offer)
-                               :instance/progress 0
-                               :instance/status :instance.status/unknown
-                               :instance/preempted? false}])
-                           task-infos)))
-              (log/info "Matched offer" offer "to tasks" task-infos)
-              ;; This launch-tasks MUST happen after the above transaction in
-              ;; order to allow a transaction failure (due to failed preconditions)
-              ;; to block the launch
-              (meters/mark! scheduler-offer-matched (count (:ids offer)))
-              (meters/mark! matched-tasks (count task-infos))
-              (histograms/update! number-tasks-matched (count task-infos))
-              (meters/mark! matched-tasks-cpus (:cpus match-resource-requirements))
-              (meters/mark! matched-tasks-mem (:mem match-resource-requirements))
-              (mesos/launch-tasks
-                driver
-                (:ids offer)
-                (mapv #(dissoc % :job-id :num-ports) task-infos))))))
+                   (vec (apply concat upgrade-txns task-txns)))
+                (log/info "Launching" (count task-txns) "tasks")
+                (log/info "Upgrading" (count (:upgrade-backfill processed-matches)) "tasks from backfilled to proper")
+                (log/debug "Matched tasks" task-txns)
+                ;; This launch-tasks MUST happen after the above transaction in
+                ;; order to allow a transaction failure (due to failed preconditions)
+                ;; to block the launch
+                (meters/mark! scheduler-offer-matched
+                              (->> matches
+                                   (mapcat (comp :id :offer :leases))
+                                   (distinct)
+                                   (count)))
+                (meters/mark! matched-tasks (count task-txns))
+                (meters/mark! matched-tasks-cpus (:cpus match-resource-requirements))
+                (meters/mark! matched-tasks-mem (:mem match-resource-requirements))
+                (doseq [{:keys [tasks leases]} matches
+                        :let [offers (mapv :offer leases)
+                              task-infos (mapv (fn process-results [^TaskAssignmentResult task]
+                                                 (reduce
+                                                   (fn add-ports-to-task-info [task-info [index port]]
+                                                     (log/debug "task-info" task-info [index port])
+                                                     (-> task-info
+                                                         (update-in [:resources :ports]
```

Comment:
Can we use the list form to store the ports that were assigned? 

Or better yet, can we have two fields in datomic:

1. num-ports-requested on job
2. ports-assigned on instance


That way we will have full information